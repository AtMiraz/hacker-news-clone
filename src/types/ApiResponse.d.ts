export type Hit = {
  created_at: string
  title: string | null
  url: string | null
  author: string
  points: string | null
  story_text: string | null
  comment_text: string | null
  num_comments: number | null
  story_id: number
  story_title: string
  story_url: string
  parent_id: number
  created_at_i: number
  _tags: string[]
  objectID: string
  _highlightResult: any
}
