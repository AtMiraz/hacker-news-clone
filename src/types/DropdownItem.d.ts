import { News } from './StoryList'

export type dropdownItem = {
  name: string
  value: News
  image: string
  selected: boolean
}
