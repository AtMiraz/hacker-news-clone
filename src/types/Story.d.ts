export type Story = {
  title: string;
  author: string;
  createdAt: string;
  url: string;
  favorite: boolean;
}