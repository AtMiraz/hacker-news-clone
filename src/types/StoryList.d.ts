export type News = 'angular' | 'reactjs' | 'vuejs'

export type StoryListProps = {
  mode?: 'all' | 'favorites'
  newsType?: News
}
