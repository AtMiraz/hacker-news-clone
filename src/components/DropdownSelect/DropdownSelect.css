.select__options-container {
  width: 15rem;
  height: 2rem;
  border-radius: 4px;
  border: solid 1px #2e2e2e;
  background-color: #fff;
}

.select__options-selected {
  padding: 5px;
  cursor: pointer;
  display: flex;
  align-items: center;
  position: relative;
}

.select__options-items {
  padding: 1px;
  border-top: 1px solid #e5e8ec;
  display: none;
}

.select__options-items.select__options-open {
  display: block;
  width: 15rem;
  border-radius: 4px;
  box-shadow: 0 2px 2px 0 #dad8d8;
}

.select__options-item {
  display: flex;
  align-items: center;
  padding: 10px;
  position: relative;
  z-index: 10;
  background-color: white;
}

.select__options-item:hover {
  cursor: pointer;
  background-color: #eaeaea;
}

.select__options-image {
  width: 1.5rem;
  height: 1.5rem;
  object-fit: contain;
}

.select__options-label {
  margin-left: 0.813rem;
}

.select__options-arrow {
  position: absolute;
  display: block;
  font-size: 0.7em;
  color: #2e2e2e;
  top: 9px;
  right: 12px;
}

.select__options-arrow.open {
  transform: rotate(-90deg);
  transition: all 0.3s ease-in;
}

.select__options-arrow.close {
  transform: rotate(0deg);
  transition: all 0.3s ease-in;
}

.select__options-arrow:before {
  content: '\25BC';
}
