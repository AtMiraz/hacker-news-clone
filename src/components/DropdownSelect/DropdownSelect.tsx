import './DropdownSelect.css'
import { useEffect, useRef, useState } from 'react'
import { handleClickOutside } from '../../utils/utils'
import { dropdownItem } from '../../types/DropdownItem'

type DropdownSelectProps = {
  items: dropdownItem[]
  onUpdate: (item: dropdownItem) => void
}

function DropdownSelect({ items, onUpdate }: DropdownSelectProps) {
  const [selectedItem, setSelectedItem] = useState<dropdownItem>()
  const [isOpen, setIsOpen] = useState(false)
  const divRef = useRef<HTMLDivElement>(null)

  function handleItemClick(item: dropdownItem) {
    setSelectedItem(item)
    onUpdate(item)
    toggleDropdown()
  }

  function toggleDropdown() {
    setIsOpen(!isOpen)
  }

  function closeDropdown() {
    setIsOpen(false)
  }

  const outsideClickHandler = handleClickOutside(
    divRef as React.MutableRefObject<HTMLDivElement>,
    closeDropdown
  )

  useEffect(() => {
    const defaultItem = items.find((item) => item.selected) || items[0]
    setSelectedItem(defaultItem)
  }, [selectedItem])

  useEffect(() => {
    if (divRef.current) {
      document.addEventListener('mousedown', outsideClickHandler)
    }
    return () => {
      document.removeEventListener('mousedown', outsideClickHandler)
    }
  }, [])

  return (
    <div
      className="select__options-container"
      ref={divRef}
      data-testid="dropdown-select"
    >
      <div
        className="select__options-selected"
        data-testid="option-selected"
        onClick={toggleDropdown}
      >
        <img
          src={selectedItem?.image}
          className="select__options-image"
          alt={selectedItem?.image}
        />
        <span data-testid="selected-option" className="select__options-label">
          {selectedItem?.name}
        </span>
        <i className={`select__options-arrow ${isOpen ? 'close' : 'open'}`}></i>
      </div>
      <div
        data-testid="select-options-items"
        className={`select__options-items ${isOpen && 'select__options-open'}`}
      >
        {items.map((item) => (
          <div
            className={`select__options-item ${
              item.selected && 'select__options-item-selected'
            }`}
            data-testid="select-options-item"
            key={item.value}
            onClick={() => handleItemClick(item)}
          >
            <img
              className="select__options-image"
              src={item.image}
              alt={item.value}
            />
            <span className="select__options-label">{item.name}</span>
          </div>
        ))}
      </div>
    </div>
  )
}

export default DropdownSelect
