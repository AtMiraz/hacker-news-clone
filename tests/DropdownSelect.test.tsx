import React from 'react'
import DropdownSelect from '../src/components/DropdownSelect/DropdownSelect'
import { fireEvent, render, screen } from '@testing-library/react'
import { expect, test, describe } from 'vitest'
import { dropdownItem } from '../src/types/DropdownItem'

describe('DropdownSelect tests', () => {
  test('should render the component', async () => {
    const component = render(
      <DropdownSelect items={[]} onUpdate={() => ({})} />
    )

    await screen.findByTestId('dropdown-select')
    expect(component).toMatchSnapshot()
    expect(component).to.exist
  })

  test('should display a selected item as default', async () => {
    const items = [
      { name: 'first entry', value: 'angular', image: '', selected: false },
      { name: 'second entry', value: 'reactjs', image: '', selected: true },
    ] as dropdownItem[]
    render(<DropdownSelect items={items} onUpdate={() => ({})} />)

    const option = await screen.findByTestId('selected-option')

    expect(option.textContent).toContain('second entry')
  })

  test('should display the items dropdown only when clicked', async () => {
    const items = [
      { name: 'first entry', value: 'angular', image: '', selected: false },
      { name: 'second entry', value: 'reactjs', image: '', selected: true },
      { name: 'second entry', value: 'vuejs', image: '', selected: false },
    ] as dropdownItem[]
    render(<DropdownSelect items={items} onUpdate={() => ({})} />)

    const isOpen = await screen.findByTestId('select-options-items')
    expect(isOpen.className).not.toContain('select__options-open')

    await fireEvent.click(await screen.findByTestId('option-selected'))

    expect(isOpen.className).toContain('select__options-open')

    const options = await screen.findAllByTestId('select-options-item')
    expect(options.length).toBe(3)
  })
})
