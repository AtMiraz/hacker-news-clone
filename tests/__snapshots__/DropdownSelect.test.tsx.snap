// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownSelect tests > should render the component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="select__options-container"
        data-testid="dropdown-select"
      >
        <div
          class="select__options-selected"
          data-testid="option-selected"
        >
          <img
            class="select__options-image"
          />
          <span
            class="select__options-label"
            data-testid="selected-option"
          />
          <i
            class="select__options-arrow open"
          />
        </div>
        <div
          class="select__options-items false"
          data-testid="select-options-items"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="select__options-container"
      data-testid="dropdown-select"
    >
      <div
        class="select__options-selected"
        data-testid="option-selected"
      >
        <img
          class="select__options-image"
        />
        <span
          class="select__options-label"
          data-testid="selected-option"
        />
        <i
          class="select__options-arrow open"
        />
      </div>
      <div
        class="select__options-items false"
        data-testid="select-options-items"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
